src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs

```diff
@@ -0,0 +1,117 @@
+using Microsoft.AspNetCore.Testing.xunit;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.FunctionalTests
+{
+    public class UnixDomainSocketsTest : TestApplicationErrorLoggerLoggedTest
+    {
+#if LIBUV
+        [OSSkipCondition(OperatingSystems.Windows, SkipReason = "Libuv does not support unix domain sockets on Windows.")]
+#else
+        [OSSkipCondition(OperatingSystems.Windows, WindowsVersions.Win7, WindowsVersions.Win8, WindowsVersions.Win81, WindowsVersions.Win2008R2, SkipReason = "UnixDomainSocketEndPoint is not supported on older versions of Windows")]
+#endif
+        [ConditionalFact]
+        public async Task TestUnixDomainSocket()
+        {
+            var path = Path.GetTempFileName();
+
+            Delete(path);
+
+            try
+            {
+                async Task EchoServer(ConnectionContext connection)
+                {
+                    // For graceful shutdown
+                    var notificationFeature = connection.Features.Get<IConnectionLifetimeNotificationFeature>();
+
+                    try
+                    {
+                        while (true)
+                        {
+                            var result = await connection.Transport.Input.ReadAsync(notificationFeature.ConnectionClosedRequested);
+
+                            if (result.IsCompleted)
+                            {
+                                break;
+                            }
+
+                            await connection.Transport.Output.WriteAsync(result.Buffer.ToArray());
+
+                            connection.Transport.Input.AdvanceTo(result.Buffer.End);
+                        }
+                    }
+                    catch (OperationCanceledException)
+                    {
+
+                    }
+                }
+
+                var hostBuilder = TransportSelector.GetWebHostBuilder()
+                    .UseKestrel(o =>
+                    {
+                        o.ListenUnixSocket(path, builder =>
+                        {
+                            builder.Run(EchoServer);
+                        });
+                    })
+                    .ConfigureServices(AddTestLogging)
+                    .Configure(c => { });
+
+                using (var host = hostBuilder.Build())
+                {
+                    await host.StartAsync();
+
+                    using (var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified))
+                    {
+                        socket.Connect(new UnixDomainSocketEndPoint(path));
+
+                        var data = Encoding.ASCII.GetBytes("Hello World");
+                        socket.Send(data);
+
+                        var buffer = new byte[data.Length];
+                        var read = 0;
+                        while (read < data.Length)
+                        {
+                            read += socket.Receive(buffer, read, buffer.Length - read, SocketFlags.None);
```

Comment:
I know it's just a test, but I'd avoid blocking. We've seen threadpool issues in the past...